src/fluree/db/reasoner/owl_datalog.cljc

```diff
@@ -526,6 +888,66 @@
                             "insert" {"@id"   "?y"
                                       "@type" (into [] class-list)}})
 
+           ;; Generate forward entailment rules for allValuesFrom in intersections
+           all-values-rules (reduce (fn [acc* all-val-restriction]
+                                      (let [{:keys [property is-inverse?]} (extract-property-with-inverse all-val-restriction)
+                                            target-class (util/get-first-id all-val-restriction const/iri-owl:allValuesFrom)]
+                                        (if (and property target-class)
+                                          (let [rule (if is-inverse?
+                                                      ;; For inverse: if x is rule-class and y has property x, then y is target-class
+                                                       {"where"  [{"@id"   "?x"
+                                                                   "@type" rule-class}
+                                                                  {"@id"    "?y"
+                                                                   property "?x"}]
+                                                        "insert" {"@id"   "?y"
+                                                                  "@type" target-class}}
+                                                      ;; Normal: if x is rule-class and x has property y, then y is target-class
+                                                       {"where"  [{"@id"   "?x"
+                                                                   "@type" rule-class}
+                                                                  {"@id"    "?x"
+                                                                   property "?y"}]
+                                                        "insert" {"@id"   "?y"
+                                                                  "@type" target-class}})]
+                                            (conj acc* [(str rule-class "(owl:allValuesFrom-forward-" property ")") rule]))
+                                          acc*)))
+                                    []
+                                    all-values)
+
+           ;; Generate forward entailment rules for hasValue in intersections
+           has-value-rules (reduce (fn [acc* has-val-restriction]
```

Comment:
I think this function would be more readable if we extract the `has-value-rules` and `all-values-rules` into their own functions.